File: 22.generate-parentheses.cpp
/*
 * @lc app=leetcode id=22 lang=cpp
 *
 * [22] Generate Parentheses
 */

// @lc code=start
class Solution {
   public:
    vector<string> generateParenthesis(int n) {
        vector<pair<string, int>> strings = {{"(", 0}};

        string s;
        int open_count;
        int close_count;
        int strings_size;

        for (int k = 1; k < 2 * n; k++) {
            strings_size = strings.size();
            for (int i = 0; i < strings_size; i++) {
                s = strings.front().first;
                close_count = strings.front().second;
                strings.erase(strings.begin());
                open_count = s.size() - close_count;

                if (open_count < n)
                    strings.push_back({s + "(", close_count});

                if (open_count > close_count)
                    strings.push_back({s + ")", close_count + 1});
            }
        }

        vector<string> result;
        for (pair<string, int> p : strings) {
            result.push_back(p.first);
        }

        return result;
    }
};
// @lc code=end

class Solution {
   public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;

        if (n == 0)
            result.push_back("");
        else {
            for (int c = 0; c < n; ++c) {
                for (string left : generateParenthesis(c))
                    for (string right : generateParenthesis(n - 1 - c))
                        result.push_back("(" + left + ")" + right);
            }
        }
    }
};

/*
class Solution {
   public:
    vector<string> generateParenthesis(int n) {
        vector<string> result = {"()"};
        if (n == 1) return result;
        vector<string> prev = generateParenthesis(n - 1);

        for (int i = 0; i < prev.size(); i++) {
            string s = prev[i];
            result.push_back("()" + s);
            result.push_back("(" + s +)

        }



        for (int i = 1; i < n; i++) {
            // printf("i: %i\n", i);
            int result_size = result.size();
            for (int j = 0; j < result_size; j++) {
                // printf("j: %i\n", j);
                string curr = result.front();

                // printf("curr: %s\n", curr.c_str());
                result.erase(result.begin());

                result.push_back("()" + curr);

                result.push_back("(" + curr + ")");

                // printf("%s\n", ("(" + curr + ")").c_str());
                // printf("%s\n", ("()" + curr).c_str());

                if (j > 0) {
                    result.push_back(curr + "()");
                    // printf("%s\n", (curr + "()").c_str());
                }
            }
        }

        return result;
    }
};

*/

File: 22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:

    """
    Brute force: create all possible strings and check if well-formed
    Time complexity: O(2^2n * n)
    Space complexity: O(2^n * n + n)
    """
#     def generateParenthesis(self, n: int) -> List[str]:
#         def generate(acc):
#             if len(acc) == 2 * n:
#                 if valid(acc):
#                     ans.append(''.join(acc))
#             else:
#                 acc.append('(')
#                 generate(acc)
#                 acc.pop()
#                 acc.append(')')
#                 generate(acc)
#                 acc.pop()
#         def valid(acc):
#             bal = 0
#             for c in acc:
#                 if c == '(':
#                     bal += 1
#                 else:
#                     bal -= 1
#                 if bal < 0:
#                     return False
#             return bal == 0

#         ans = []
#         generate([])
#         return ans

    """
    Closure Number
    """

    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return ['']
        ans = []
        for c in range(n):
            for left in self.generateParenthesis(c):
                for right in self.generateParenthesis(n-1-c):
                    ans.append('({}){}'.format(left, right))
        return ans

    """
    Brute force: create all possible strings and check if well-formed
    Time complexity: O(2^n)
    Space complexity: O(2^n * n + n ^ 2)
    """
#     def generateParenthesis(self, n: int) -> List[str]:
#         results = []

#         def rec(s, open_count, close_count):
#             if open_count < close_count or open_count > n:
#                 return
#             if len(s) == n * 2:
#                 results.append(s)
#                 return

#             rec(s + '(', open_count + 1, close_count)
#             rec(s + ')', open_count, close_count + 1)

#         rec('', 0, 0)
#         return results

    """
    Backtracking (Stack)
    Time complexity: O(# results * n) -> # 
    Space complexity: O(# results * n + n)
    """
#     def generateParenthesis(self, n: int) -> List[str]:
#         results = []
#         stack = []
#         final_len = 2 * n
#         def build(open_count, close_count):
#             if len(stack) == final_len:
#                 results.append(''.join(stack))
#                 return
#             if open_count - close_count > 0:
#                 stack.append(')')
#                 build(open_count, close_count + 1)
#                 stack.pop()
#             if open_count < n:
#                 stack.append('(')
#                 build(open_count + 1, close_count)
#                 stack.pop()

#         build(0, 0)
#         return results


# @lc code=end
